js.go

```
@@ -846,21 +846,228 @@ func (sub *Subscription) ConsumerInfo() (*ConsumerInfo, error) {
+	// - 409 in case it reached MaxAckPending.
+	//
+	// PullExpiresIn will give a duration to the server to expire our pull request
+	// to this inbox only and will serve the newest one.
+	//
+	// Default is to let the pull requests linger until there are 512 waiting requests,
+	// this limit can be modified by using PullMaxWaiting when creating the consumer.
+	//
+	nr := &nextRequest{Batch: batch, NoWait: o.noWait, Expires: o.expires}
+	req, _ := json.Marshal(nr)
 	reqNext := js.apiSubj(fmt.Sprintf(apiRequestNextT, stream, consumer))
-	return nc.PublishRequest(reqNext, reply, req)
+
+	// Make an old style request, we will wait for first response
+	// in case of errors, then dispatch the rest of the replies
+	// to the channel.
+	inbox := NewInbox()
+
+	mch := make(chan *Msg, batch)
+	s, err := nc.subscribe(inbox, _EMPTY_, nil, mch, true, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Remove interest in the subscription at the end so that the
+	// this inbox does not get delivered the results intended
+	// for another request.
+	defer s.Unsubscribe()
+
+	s.AutoUnsubscribe(batch)
+
+	// Make a publish request to get results of the pull.
+	err = nc.publish(reqNext, inbox, nil, req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check for empty payload message and process synchronously
+	// any status messages.
+	checkMsg := func(msg *Msg) error {
+		if len(msg.Data) == 0 {
+			switch msg.Header.Get(statusHdr) {
+			case noResponders:
+				return ErrNoResponders
+			case "400", "404", "408", "409":
+				return errors.New(msg.Header.Get(descrHdr))
+			}
+		}
+		return nil
+	}
+
+	msgs := make([]*Msg, 0)
+
+	select {
+	case msg, ok := <-mch:
+		if !ok {
+			err = s.getNextMsgErr()
+		} else {
+			err = s.processNextMsgDelivered(msg)
+			if err == nil {
+				err = checkMsg(msg)
+			}
+		}
+		msgs = append(msgs, msg)
+	case <-ctx.Done():
+		if ctx.Err() == context.Canceled {
+			err = ctx.Err()
+		} else {
+			err = ErrTimeout
```

@ColinSullivan1 If we don't get a single message, should we return a timeout error or an empty collection of messages? wdyt?

---

I'd think a timeout would be consistent with the rest of the API, but am interested in other's opinions.  Polls typically wait for data and timeout when there's no response.

---

In case both the no wait first request and longer request both fail, then it is an error. Otherwise the set of delivered messages is returned.